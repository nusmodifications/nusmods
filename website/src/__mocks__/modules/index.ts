import { Module } from 'types/modules';

import ACC2002_JSON from './ACC2002.json';
import BFS1001_JSON from './BFS1001.json';
import CS1010S_JSON from './CS1010S.json';
import CS3216_JSON from './CS3216.json';
import CS4243_JSON from './CS4243.json';
import GES1021_JSON from './GES1021.json';
import PC1222_JSON from './PC1222.json';

// Have to cast these as Module explicitly, otherwise TS will try to
// incorrectly infer the shape from the JSON - specifically Weeks will
// not be cast correctly
export const CS1010S: Module = CS1010S_JSON;
export const ACC2002: Module = ACC2002_JSON;
export const BFS1001: Module = BFS1001_JSON;
export const CS3216: Module = CS3216_JSON;
export const GES1021: Module = GES1021_JSON;
export const PC1222: Module = PC1222_JSON;
export const CS4243: Module = CS4243_JSON;

const modules: Module[] = [ACC2002, BFS1001, CS1010S, CS3216, GES1021, PC1222];
export default modules;
