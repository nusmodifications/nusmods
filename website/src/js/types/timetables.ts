import { ClassNo, LessonType, ModuleCode, ModuleTitle, RawLesson } from './modules';

//  ModuleLessonConfig is a mapping of lessonType to ClassNo for a module.
export type ModuleLessonConfig = {
  [lessonType: string]: ClassNo;
};

// SemTimetableConfig is the timetable data for each semester.
export type SemTimetableConfig = {
  [moduleCode: string]: ModuleLessonConfig;
};

//  ModuleLessonConfigWithLessons is a mapping of lessonType to an array of Lessons for a module.
export type Lesson = RawLesson & {
  moduleCode: ModuleCode;
  title: ModuleTitle;
};

export type ColoredLesson = Lesson & {
  colorIndex: ColorIndex;
};

type Modifiable = {
  isModifiable?: boolean;
  isAvailable?: boolean;
  isActive?: boolean;
  colorIndex: ColorIndex;
};

export type ModifiableLesson = ColoredLesson & Modifiable;
//  The array of Lessons must belong to that lessonType.
export type ModuleLessonConfigWithLessons = {
  [lessonType: string]: Lesson[];
};

// SemTimetableConfig is the timetable data for each semester with lessons data.
export type SemTimetableConfigWithLessons = {
  [moduleCode: string]: ModuleLessonConfigWithLessons;
};

// TimetableConfig is the timetable data for the whole academic year.
export type TimetableConfig = {
  [semester: string]: SemTimetableConfig;
};

// TimetableDayFormat is timetable data grouped by DayText.
export type TimetableDayFormat = {
  [dayText: string]: ColoredLesson[];
};

// TimetableDayArrangement is the arrangement of lessons on the timetable within a day.
export type TimetableDayArrangement = ModifiableLesson[][];

// TimetableArrangement is the arrangement of lessons on the timetable for a week.
export type TimetableArrangement = {
  [dayText: string]: TimetableDayArrangement;
};

// Represents the lesson which the user is currently hovering over.
// Used to highlight lessons which have the same classNo
export type HoverLesson = {
  readonly classNo: ClassNo;
  readonly moduleCode: ModuleCode;
  readonly lessonType: LessonType;
};

export type ColorIndex = number;
