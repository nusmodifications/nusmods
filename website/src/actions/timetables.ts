import { each, flatMap } from 'lodash';

import type { Lesson, ColorIndex, ModuleLessonConfig, SemTimetableConfig } from 'types/timetables';
import type { Dispatch, GetState } from 'types/redux';
import type { ColorMapping } from 'types/reducers';
import type { ClassNo, LessonType, Module, ModuleCode, Semester } from 'types/modules';

import { fetchModule } from 'actions/moduleBank';
import { openNotification } from 'actions/app';
import { getModuleCondensed } from 'selectors/moduleBank';
import {
  randomModuleLessonConfig,
  validateModuleLessons,
  validateTimetableModules,
} from 'utils/timetables';
import { getModuleTimetable } from 'utils/modules';

// Actions that should not be used directly outside of thunks
export const SET_TIMETABLE = 'SET_TIMETABLE' as const;
export const ADD_MODULE = 'ADD_MODULE' as const;
export const Internal = {
  setTimetable(
    semester: Semester,
    timetable: SemTimetableConfig | undefined,
    colors?: ColorMapping,
  ) {
    return {
      type: SET_TIMETABLE,
      payload: { semester, timetable, colors },
    };
  },

  addModule(semester: Semester, moduleCode: ModuleCode, moduleLessonConfig: ModuleLessonConfig) {
    return {
      type: ADD_MODULE,
      payload: {
        semester,
        moduleCode,
        moduleLessonConfig,
      },
    };
  },
};

export function addModule(semester: Semester, moduleCode: ModuleCode) {
  return (dispatch: Dispatch, getState: GetState) =>
    dispatch(fetchModule(moduleCode)).then(() => {
      const module: Module = getState().moduleBank.modules[moduleCode];

      if (!module) {
        dispatch(
          openNotification(`Cannot load ${moduleCode}`, {
            action: {
              text: 'Retry',
              handler: () => {
                dispatch(addModule(semester, moduleCode));
              },
            },
          }),
        );

        return;
      }

      const lessons = getModuleTimetable(module, semester);
      const moduleLessonConfig = randomModuleLessonConfig(lessons);

      dispatch(Internal.addModule(semester, moduleCode, moduleLessonConfig));
    });
}

export const REMOVE_MODULE = 'REMOVE_MODULE' as const;
export function removeModule(semester: Semester, moduleCode: ModuleCode) {
  return {
    type: REMOVE_MODULE,
    payload: {
      semester,
      moduleCode,
    },
  };
}

export const MODIFY_LESSON = 'MODIFY_LESSON' as const;
export function modifyLesson(activeLesson: Lesson) {
  return {
    type: MODIFY_LESSON,
    payload: {
      activeLesson,
    },
  };
}

export const CHANGE_LESSON = 'CHANGE_LESSON' as const;
export function setLesson(
  semester: Semester,
  moduleCode: ModuleCode,
  lessonType: LessonType,
  classNo: ClassNo,
) {
  return {
    type: CHANGE_LESSON,
    payload: {
      semester,
      moduleCode,
      lessonType,
      classNo,
    },
  };
}

export function changeLesson(semester: Semester, lesson: Lesson) {
  return setLesson(semester, lesson.moduleCode, lesson.lessonType, lesson.classNo);
}

export const SET_LESSON_CONFIG = 'SET_LESSON_CONFIG' as const;
export function setLessonConfig(
  semester: Semester,
  moduleCode: ModuleCode,
  lessonConfig: ModuleLessonConfig,
) {
  return {
    type: SET_LESSON_CONFIG,
    payload: {
      semester,
      moduleCode,
      lessonConfig,
    },
  };
}

export const CANCEL_MODIFY_LESSON = 'CANCEL_MODIFY_LESSON' as const;
export function cancelModifyLesson() {
  return {
    type: CANCEL_MODIFY_LESSON,
    payload: null,
  };
}

export function setTimetable(
  semester: Semester,
  timetable?: SemTimetableConfig,
  colors?: ColorMapping,
) {
  return (dispatch: Dispatch, getState: GetState) => {
    let validatedTimetable = timetable;
    if (timetable) {
      [validatedTimetable] = validateTimetableModules(timetable, getState().moduleBank.moduleCodes);
    }

    return dispatch(Internal.setTimetable(semester, validatedTimetable, colors));
  };
}

export function validateTimetable(semester: Semester) {
  return (dispatch: Dispatch, getState: GetState) => {
    const { timetables, moduleBank } = getState();

    // Extract the timetable and the modules for the semester
    const timetable = timetables.lessons[semester];
    if (!timetable) return;

    // Check that all lessons for each module are valid. If they are not, we update it
    // such that they are
    each(timetable, (lessonConfig: ModuleLessonConfig, moduleCode: ModuleCode) => {
      const module = moduleBank.modules[moduleCode];
      if (!module) return;

      const [validatedLessonConfig, changedLessonTypes] = validateModuleLessons(
        semester,
        lessonConfig,
        module,
      );

      if (changedLessonTypes.length) {
        dispatch(setLessonConfig(semester, moduleCode, validatedLessonConfig));
      }
    });
  };
}

export function fetchTimetableModules(timetables: SemTimetableConfig[]) {
  return (dispatch: Dispatch, getState: GetState) => {
    const moduleCodes = new Set(flatMap(timetables, Object.keys));
    const validateModule = getModuleCondensed(getState());

    return Promise.all(
      Array.from(moduleCodes)
        .filter(validateModule)
        .map((moduleCode) => dispatch(fetchModule(moduleCode))),
    );
  };
}

export const SELECT_MODULE_COLOR = 'SELECT_MODULE_COLOR' as const;
export function selectModuleColor(
  semester: Semester,
  moduleCode: ModuleCode,
  colorIndex: ColorIndex,
) {
  return {
    type: SELECT_MODULE_COLOR,
    payload: {
      semester,
      moduleCode,
      colorIndex,
    },
  };
}

export const HIDE_LESSON_IN_TIMETABLE = 'HIDE_LESSON_IN_TIMETABLE' as const;
export function hideLessonInTimetable(semester: Semester, moduleCode: ModuleCode) {
  return {
    type: HIDE_LESSON_IN_TIMETABLE,
    payload: { moduleCode, semester },
  };
}

export const SHOW_LESSON_IN_TIMETABLE = 'SHOW_LESSON_IN_TIMETABLE' as const;
export function showLessonInTimetable(semester: Semester, moduleCode: ModuleCode) {
  return {
    type: SHOW_LESSON_IN_TIMETABLE,
    payload: { moduleCode, semester },
  };
}
