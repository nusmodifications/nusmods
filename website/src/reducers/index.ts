import { REMOVE_MODULE, SET_TIMETABLE } from 'actions/timetables';

import persistReducer from 'storage/persistReducer';
import { State } from 'types/state';
import { Actions } from 'types/actions';

// Non-persisted reducers
import requests from './requests';
import app from './app';
import createUndoReducer from './undoHistory';

// Persisted reducers
import moduleBankReducer, { persistConfig as moduleBankPersistConfig } from './moduleBank';
import venueBankReducer, { persistConfig as venueBankPersistConfig } from './venueBank';
import timetablesReducer, { persistConfig as timetablesPersistConfig } from './timetables';
import themeReducer from './theme';
import settingsReducer, { persistConfig as settingsPersistConfig } from './settings';
import plannerReducer, { persistConfig as plannerPersistConfig } from './planner';

// Persist reducers
const moduleBank = persistReducer('moduleBank', moduleBankReducer, moduleBankPersistConfig);
const venueBank = persistReducer('venueBank', venueBankReducer, venueBankPersistConfig);
const timetables = persistReducer('timetables', timetablesReducer, timetablesPersistConfig);
const theme = persistReducer('theme', themeReducer);
const settings = persistReducer('settings', settingsReducer, settingsPersistConfig);
const planner = persistReducer('planner', plannerReducer, plannerPersistConfig);

// State default is delegated to its child reducers.
const defaultState = {} as unknown as State;
const undoReducer = createUndoReducer<State>({
  limit: 1,
  actionsToWatch: [REMOVE_MODULE, SET_TIMETABLE],
  storedKeyPaths: ['timetables', 'theme.colors'],
});

export default function reducers(state: State = defaultState, action: Actions): State {
  // Update every reducer except the undo reducer
  const newState: State = {
    moduleBank: moduleBank(state.moduleBank, action),
    venueBank: venueBank(state.venueBank, action),
    requests: requests(state.requests, action),
    timetables: timetables(state.timetables, action),
    app: app(state.app, action),
    theme: theme(state.theme, action),
    settings: settings(state.settings, action),
    planner: planner(state.planner, action),
    undoHistory: state.undoHistory,
  };
  return undoReducer(state, newState, action);
}
