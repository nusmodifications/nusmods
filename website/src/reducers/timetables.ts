import { get, omit, values } from 'lodash';
import produce from 'immer';
import { createMigrate } from 'redux-persist';

import { PersistConfig } from 'storage/persistReducer';
import { FSA } from 'types/redux';
import { ModuleCode, Semester } from 'types/modules';
import { ModuleLessonConfig, SemTimetableConfig } from 'types/timetables';
import { ColorMapping, TimetablesState } from 'types/reducers';

import config from 'config';
import {
  ADD_MODULE,
  CHANGE_LESSON,
  HIDE_LESSON_IN_TIMETABLE,
  REMOVE_MODULE,
  SELECT_MODULE_COLOR,
  SET_LESSON_CONFIG,
  SET_TIMETABLE,
  SHOW_LESSON_IN_TIMETABLE,
} from 'actions/timetables';
import { getNewColor } from 'utils/colors';
import { SET_EXPORTED_DATA } from 'actions/constants';

const EMPTY_OBJECT = {};

export const persistConfig = {
  /* eslint-disable no-useless-computed-key */
  migrate: createMigrate({
    [1]: (state) => ({
      ...state,
      archive: {},
    }),
  }),
  /* eslint-enable */
  version: 1,

  // Our own state reconciler archives old timetables if the acad year is different,
  // otherwise use the persisted timetable state
  stateReconciler: (
    inbound: TimetablesState,
    original: TimetablesState,
    reduced: TimetablesState,
    { debug }: PersistConfig,
  ): TimetablesState => {
    if (inbound.academicYear === original.academicYear) {
      return inbound;
    }

    if (debug) {
      // eslint-disable-next-line no-console
      console.log(
        'New academic year detected - resetting timetable and adding timetable to archive',
      );
    }

    return {
      ...original,
      archive: {
        ...inbound.archive,
        [inbound.academicYear]: inbound.lessons,
      },
    };
  },
};

// Map of lessonType to ClassNo.
const defaultModuleLessonConfig: ModuleLessonConfig = {};

function moduleLessonConfig(
  state: ModuleLessonConfig = defaultModuleLessonConfig,
  action: FSA,
): ModuleLessonConfig {
  if (!action.payload) return state;

  switch (action.type) {
    case CHANGE_LESSON: {
      const { classNo, lessonType } = action.payload;
      if (!(classNo && lessonType)) return state;
      return {
        ...state,
        [lessonType]: classNo,
      };
    }
    case SET_LESSON_CONFIG:
      return action.payload.lessonConfig;

    default:
      return state;
  }
}

// Map of ModuleCode to module lesson config.
const defaultSemTimetableConfig: SemTimetableConfig = {};
function semTimetable(
  state: SemTimetableConfig = defaultSemTimetableConfig,
  action: FSA,
): SemTimetableConfig {
  const moduleCode = get(action, 'payload.moduleCode');
  if (!moduleCode) return state;

  switch (action.type) {
    case ADD_MODULE:
      return {
        ...state,
        [moduleCode]: action.payload.moduleLessonConfig,
      };
    case REMOVE_MODULE:
      return omit(state, [moduleCode]);
    case CHANGE_LESSON:
    case SET_LESSON_CONFIG:
      return {
        ...state,
        [moduleCode]: moduleLessonConfig(state[moduleCode], action),
      };
    default:
      return state;
  }
}

// Map of semester to color mapping
const defaultSemColorMap = {};
function semColors(state: ColorMapping = defaultSemColorMap, action: FSA): ColorMapping {
  const moduleCode = get(action, 'payload.moduleCode');
  if (!moduleCode) return state;

  switch (action.type) {
    case ADD_MODULE:
      return {
        ...state,
        [moduleCode]: getNewColor(values(state)),
      };

    case REMOVE_MODULE:
      return omit(state, moduleCode);

    case SELECT_MODULE_COLOR:
      return {
        ...state,
        [moduleCode]: action.payload.colorIndex,
      };

    default:
      return state;
  }
}

// Map of semester to list of hidden modules
const defaultHiddenState: ModuleCode[] = [];
function semHiddenModules(state = defaultHiddenState, action: FSA) {
  if (!action.payload) {
    return state;
  }

  switch (action.type) {
    case HIDE_LESSON_IN_TIMETABLE:
      return [action.payload.moduleCode, ...state];
    case SHOW_LESSON_IN_TIMETABLE:
    case REMOVE_MODULE:
      return state.filter((c) => c !== action.payload.moduleCode);
    default:
      return state;
  }
}

export const defaultTimetableState: TimetablesState = {
  lessons: {},
  colors: {},
  hidden: {},
  academicYear: config.academicYear,
  archive: {},
};

function timetables(state: TimetablesState = defaultTimetableState, action: FSA): TimetablesState {
  // All normal timetable actions should specify their semester
  if (!action.payload || !action.payload.semester) {
    return state;
  }

  switch (action.type) {
    case SET_TIMETABLE: {
      const { semester, timetable, colors } = action.payload;

      return produce(state, (draft) => {
        draft.lessons[semester] = timetable || defaultSemTimetableConfig;
        draft.colors[semester] = colors || {};
      });
    }

    case ADD_MODULE:
    case REMOVE_MODULE:
    case SELECT_MODULE_COLOR:
    case CHANGE_LESSON:
    case SET_LESSON_CONFIG:
    case HIDE_LESSON_IN_TIMETABLE:
    case SHOW_LESSON_IN_TIMETABLE: {
      const { semester } = action.payload;

      return produce(state, (draft) => {
        draft.lessons[semester] = semTimetable(draft.lessons[semester], action);
        draft.colors[semester] = semColors(state.colors[semester], action);
        draft.hidden[semester] = semHiddenModules(state.hidden[semester], action);
      });
    }

    case SET_EXPORTED_DATA: {
      const { semester, timetable, colors, hidden } = action.payload;

      return {
        ...state,
        lessons: { [semester]: timetable },
        colors: { [semester]: colors },
        hidden: { [semester]: hidden },
      };
    }

    default:
      return state;
  }
}

export default timetables;

// Extract sem timetable and colors for a specific semester from TimetablesState
export function getSemesterTimetable(
  semester: Semester,
  state: TimetablesState,
): { timetable: SemTimetableConfig; colors: ColorMapping } {
  return {
    timetable: state.lessons[semester] || EMPTY_OBJECT,
    colors: state.colors[semester] || EMPTY_OBJECT,
  };
}
