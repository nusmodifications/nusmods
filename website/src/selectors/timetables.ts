import { ModuleCode, Semester } from 'types/modules';
import config from 'config';
import { isOngoing, isSuccess } from 'selectors/requests';
import { State } from 'types/state';
import { fetchArchiveRequest } from 'actions/constants';
import { ColorMapping, TimetablesState } from 'types/reducers';
import { SemTimetableConfig } from 'types/timetables';

export function isArchiveLoading(state: State, moduleCode: ModuleCode) {
  return config.archiveYears.some((year) =>
    isOngoing(state, fetchArchiveRequest(moduleCode, year)),
  );
}

export function availableArchive(state: State, moduleCode: ModuleCode): string[] {
  return config.archiveYears.filter((year) =>
    isSuccess(state, fetchArchiveRequest(moduleCode, year)),
  );
}

// Extract sem timetable and colors for a specific semester from TimetablesState
const EMPTY_OBJECT = {};
export function getSemesterTimetable(
  semester: Semester,
  state: TimetablesState,
): { timetable: SemTimetableConfig; colors: ColorMapping } {
  return {
    timetable: state.lessons[semester] || EMPTY_OBJECT,
    colors: state.colors[semester] || EMPTY_OBJECT,
  };
}
