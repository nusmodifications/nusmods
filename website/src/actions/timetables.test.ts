import { ModuleCode, Semester } from 'types/modules';
import {
  SemTimetableConfig,
  LessonWithIndex,
  TaModulesConfig,
  ClassNoTimetableConfig,
} from 'types/timetables';

import lessons from '__mocks__/lessons-array.json';
import { CS1010A, CS1010S, CS3216 } from '__mocks__/modules';

import { ClassNoTaModulesMap, ModuleBank, TimetablesState } from 'types/reducers';
import { defaultTimetableState } from 'reducers/timetables';
import * as actions from './timetables';

const initialState = defaultTimetableState;

jest.mock('storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
}));

// see: https://github.com/reactjs/redux/blob/master/docs/recipes/WritingTests.md#example-1
// TODO: write addModule test with nock and mockStore.
test('addModule should create an action to add a module', () => {
  const moduleCode = 'CS1010';
  const semester = 1;

  const value = actions.addModule(semester, moduleCode);
  // TODO
  expect(value).toBeInstanceOf(Function);
});

test('removeLesson should return information to remove module', () => {
  const semester: Semester = 1;
  const moduleCode: ModuleCode = 'CS1010';
  expect(actions.removeModule(semester, moduleCode)).toMatchSnapshot();
});

test('modifyLesson should return lesson payload', () => {
  const activeLesson: LessonWithIndex = lessons[0];
  expect(actions.modifyLesson(activeLesson)).toMatchSnapshot();
});

test('changeLesson should return updated information to change lesson', () => {
  const semester: Semester = 1;
  const lesson: LessonWithIndex = lessons[1];
  expect(
    actions.changeLesson(semester, lesson.moduleCode, lesson.lessonType, [lesson.lessonIndex]),
  ).toMatchSnapshot();
});

test('cancelModifyLesson should not have payload', () => {
  expect(actions.cancelModifyLesson()).toMatchSnapshot();
});

test('select module color should dispatch a select of module color', () => {
  const semester: Semester = 1;
  expect(actions.selectModuleColor(semester, 'CS1010S', 0)).toMatchSnapshot();
  expect(actions.selectModuleColor(semester, 'CS3216', 1)).toMatchSnapshot();
});

describe('disableTaMode', () => {
  const semester = 1;
  const timetablesState = (ta: TaModulesConfig): TimetablesState => ({
    ...initialState,
    lessons: {
      [semester]: {
        CS1010S: {
          Lecture: [0],
          Tutorial: [11],
          Recitation: [1],
        },
      },
    },
    ta: { [semester]: ta },
  });

  test('TA module is removed correctly', () => {
    const ta = ['CS1010S'];

    const state: any = {
      timetables: timetablesState(ta),
      moduleBank: { modules: { CS1010S, CS3216 } },
    };
    const dispatch = jest.fn();
    const action = actions.disableTaModule(semester, 'CS1010S');

    action(dispatch, () => state);

    expect(dispatch).toHaveBeenCalled();
  });

  test('TA module is removed even if semesterData cannot be found to create normal mode lessonConfig', () => {
    const ta = ['CS1010S'];

    const state: any = {
      timetables: timetablesState(ta),
      moduleBank: { modules: { CS1010S: { semesterData: [] } } },
    };
    const dispatch = jest.fn();
    const action = actions.disableTaModule(semester, 'CS1010S');

    action(dispatch, () => state);

    expect(dispatch).toHaveBeenCalled();
  });
});

describe('fillTimetableBlanks', () => {
  const moduleBank: Partial<ModuleBank> = { modules: { CS1010S, CS1010A, CS3216 } };
  const semester: Semester = 1;
  const timetablesState = (timetable: SemTimetableConfig): TimetablesState => ({
    ...initialState,
    lessons: { [semester]: timetable },
  });
  const action = actions.validateTimetable(semester);

  test('do nothing if timetable is already full', () => {
    const timetable = {
      CS1010S: {
        Lecture: [0],
        Tutorial: [11],
        Recitation: [1],
      },
    };

    const state: any = { timetables: timetablesState(timetable), moduleBank };
    const dispatch = jest.fn();
    action(dispatch, () => state);

    expect(dispatch).not.toHaveBeenCalled();
  });

  test('fill missing lessons with randomly generated modules', () => {
    const timetable = {
      CS1010S: {
        Lecture: [0],
        Tutorial: [11],
      },
      CS3216: {},
    };
    const state: any = { timetables: timetablesState(timetable), moduleBank };
    const dispatch = jest.fn();

    action(dispatch, () => state);

    expect(dispatch).toHaveBeenCalledTimes(2);

    const [[firstAction], [secondAction]] = dispatch.mock.calls;
    expect(firstAction).toMatchObject({
      type: actions.SET_LESSON_CONFIG,
      payload: {
        semester,
        moduleCode: 'CS1010S',
        lessonConfig: {
          Lecture: [0],
          Tutorial: [11],
          Recitation: [1],
        },
      },
    });

    expect(secondAction).toMatchObject({
      type: actions.SET_LESSON_CONFIG,
      payload: {
        semester,
        moduleCode: 'CS3216',
        lessonConfig: {
          Lecture: [0],
        },
      },
    });
  });

  test('migrate v1 config', () => {
    const timetables = {
      ...initialState,
      lessons: {
        [semester]: {
          CS1010S: {
            Lecture: '1',
            Tutorial: '1',
            Recitation: '1',
          },
          CS3216: {
            Lecture: '1',
          },
        } as ClassNoTimetableConfig,
      },
      ta: {
        [semester]: {
          CS1010S: [
            ['Lecture', '1'],
            ['Tutorial', '2'],
            ['Recitation', '2'],
          ],
        },
      } as ClassNoTaModulesMap,
    };

    const state: any = { timetables, moduleBank };
    const dispatch = jest.fn();
    action(dispatch, () => state);
    expect(dispatch).toHaveBeenCalledTimes(1);
    const [[firstAction]] = dispatch.mock.calls;
    expect(firstAction).toMatchObject({
      type: 'SET_TIMETABLES',
      payload: {
        lessons: {
          [semester]: {
            CS1010S: {
              Lecture: [0],
              Tutorial: [21],
              Recitation: [3],
            },
          },
        },
        taModules: {
          [semester]: ['CS1010S'],
        },
      },
    });
  });

  test('should not error when module cannot be found', () => {
    const timetable = {
      CS1010S: {
        Lecture: [0],
        Tutorial: [11],
        Recitation: [1],
      },
    };
    const moduleBankWithoutModule = {
      ...moduleBank,
      modules: {},
    };

    const state: any = {
      timetables: timetablesState(timetable),
      moduleBank: moduleBankWithoutModule,
    };
    const dispatch = jest.fn();
    action(dispatch, () => state);

    expect(dispatch).not.toThrow(TypeError);
  });
});

describe('hide/show timetable modules', () => {
  const semester: Semester = 1;

  test('should dispatch a module code for hiding', () => {
    const moduleCode: ModuleCode = 'CS1010';
    expect(actions.hideLessonInTimetable(semester, moduleCode)).toMatchSnapshot();
  });

  test('should dispatch a module code for showing', () => {
    const moduleCode: ModuleCode = 'CS1020';
    expect(actions.showLessonInTimetable(semester, moduleCode)).toMatchSnapshot();
  });
});

describe(actions.fetchTimetableModules, () => {
  const moduleCodes: any = {
    CS1010S: {},
  };

  const state: any = {
    moduleBank: {
      moduleCodes,
    },
  };

  const dispatch = jest.fn().mockResolvedValue(undefined);
  const getState = jest.fn().mockReturnValue(state);

  beforeEach(() => {
    dispatch.mockClear();
    getState.mockClear();
  });

  test('should fetch modules', async () => {
    const timetable = {
      CS1010S: {},
    };

    const thunk = actions.fetchTimetableModules([timetable]);
    expect(thunk).toBeInstanceOf(Function);

    await thunk(dispatch, getState);

    expect(dispatch).toBeCalled();
    expect(dispatch.mock.calls).toMatchSnapshot();
  });

  test('should not fetch invalid modules', async () => {
    const timetable = {
      invalid: {},
    };

    const thunk = actions.fetchTimetableModules([timetable]);
    await thunk(dispatch, getState);
    expect(dispatch).not.toBeCalled();
  });
});
