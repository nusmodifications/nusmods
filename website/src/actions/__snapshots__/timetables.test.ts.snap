// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cancelModifyLesson should not have payload 1`] = `
{
  "payload": null,
  "type": "CANCEL_MODIFY_LESSON",
}
`;

exports[`changeLesson should return updated information to change lesson 1`] = `
{
  "payload": {
    "classNo": "1",
    "lessonType": "Recitation",
    "moduleCode": "CS1010S",
    "semester": 1,
  },
  "type": "CHANGE_LESSON",
}
`;

exports[`fetchTimetableModules should fetch modules 1`] = `
[
  [
    [Function],
  ],
]
`;

exports[`hide/show timetable modules should dispatch a module code for hiding 1`] = `
{
  "payload": {
    "moduleCode": "CS1010",
    "semester": 1,
  },
  "type": "HIDE_LESSON_IN_TIMETABLE",
}
`;

exports[`hide/show timetable modules should dispatch a module code for showing 1`] = `
{
  "payload": {
    "moduleCode": "CS1020",
    "semester": 1,
  },
  "type": "SHOW_LESSON_IN_TIMETABLE",
}
`;

exports[`modifyLesson should return lesson payload 1`] = `
{
  "payload": {
    "activeLesson": {
      "classNo": "1",
      "colorIndex": 0,
      "day": "Wednesday",
      "endTime": "1200",
      "lessonType": "Lecture",
      "moduleCode": "CS1010S",
      "startTime": "1000",
      "title": "Programming Methodology",
      "venue": "LT26",
      "weeks": [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
      ],
    },
  },
  "type": "MODIFY_LESSON",
}
`;

exports[`removeLesson should return information to remove module 1`] = `
{
  "payload": {
    "moduleCode": "CS1010",
    "semester": 1,
  },
  "type": "REMOVE_MODULE",
}
`;

exports[`select module color should dispatch a select of module color 1`] = `
{
  "payload": {
    "colorIndex": 0,
    "moduleCode": "CS1010S",
    "semester": 1,
  },
  "type": "SELECT_MODULE_COLOR",
}
`;

exports[`select module color should dispatch a select of module color 2`] = `
{
  "payload": {
    "colorIndex": 1,
    "moduleCode": "CS3216",
    "semester": 1,
  },
  "type": "SELECT_MODULE_COLOR",
}
`;
